### PR TITLE
Hand roll UTF-16 conversion in CGUITTFont

### LINKED ISSUE #13990: FPS drops when debug information is shown (Windows MSVC build)
### Minetest version

```true
minetest 5.8.0-dev-8vf6e00
```

### Active renderer

OpenGL 4.6.0

### Irrlicht device

_No response_

### Operating system and version

Windows 10 Home 22H2

### CPU model

11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz   2.42 GHz

### GPU model

_No response_

### OpenGL version

4.6.0

### Summary

When I press f5 to open the debug menu, the fps first says about 60fps but immediately decreases to about 40 to 32 fps. This was not the case in previous development builds. I have confirmed that it happens in both single player and multiplayer when running the default minetest game with no mods enabled.

### Steps to reproduce

You should be able to replicate it by opening the debug menu.

### PR DESCRIPTION
**Pull request by sfan5**
fixes #13990

this isn't even the first hand-rolled impl we have, compare to here: https://github.com/minetest/minetest/blob/5d3e83017679317c27fe02b7087effd9d67f79cc/src/network/networkpacket.cpp#L166-L177

### CODE REVIEW

**Comment by sfan5** [comment]
@srifqi please test

**Comment by srifqi** [comment]
The FPS drop does not exist any more. I do not understand the conversion code, though.

**Comment by sfan5** [comment]
Don't worry, it just needs to work :wink:

**Review by grorp — APPROVED** [review]
I have no idea whether the code is correct either, but it seems to work and it fixes the FPS drop issue.

The MinGW build seems to dislike emojis (both before and after this PR), if I try to paste them, it says `Some exception: "wstring_convert::to_bytes"`.

On MSVC before this PR, removing an emoji using backspace results in `Some exception: "bad conversion"`. After this PR, it doesn't happen anymore, so that's a plus. Of course, the emojis aren't rendered.

So to sum up, nothing seems to be more broken than before.

**Comment by sfan5** [comment]
> The MinGW build seems to dislike emojis (both before and after this PR), if I try to paste them, it says `Some exception: "wstring_convert::to_bytes"`.

Interesting, this could be a regression from when we started using that in irrlicht https://github.com/minetest/irrlicht/commit/c766c3a023e52a31a0a0aeee1780e0000aabaec4

**Comment by grorp** [comment]
> Interesting, this could be a regression from when we started using that in irrlicht https://github.com/minetest/irrlicht/commit/c766c3a023e52a31a0a0aeee1780e0000aabaec4

Yeah
It would be interesting to test this with CIrrDeviceSDL on Windows, that would probably give completely different results.

File: src/irrlicht_changes/CGUITTFont.cpp [hunk 1]
```diff
@@ -1251,21 +1251,31 @@ std::u32string CGUITTFont::convertWCharToU32String(const wchar_t* const charArra
...
 	static_assert(sizeof(wchar_t) == 2 || sizeof(wchar_t) == 4, "unexpected wchar size");
 
-	if (sizeof(wchar_t) == 4) // Systems where wchar_t is UTF-32
+	if (sizeof(wchar_t) == 4) // wchar_t is UTF-32
 		return std::u32string(reinterpret_cast<const char32_t*>(charArray));
 
-	// Systems where wchar_t is UTF-16:
-	// First, convert to UTF-8
-	std::u16string utf16String(reinterpret_cast<const char16_t*>(charArray));
-	std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t> converter1;
```
Comment (SmallJoker) [inline src/irrlicht_changes/CGUITTFont.cpp:1260]: `#include <codecvt>` is no longer necessary. The code still compiles.

Comment (grorp) [reply src/irrlicht_changes/CGUITTFont.cpp:1260]: `<locale>` was also added by #13775.
